### PR TITLE
tstest/natlab: add a configurable SNAT44 translator.

### PR DESCRIPTION
**Pull request by danderson**
This lets us implement the most common kinds of NAT in the wild.

Signed-off-by: David Anderson <danderson@tailscale.com>

---
This change is [Reviewable](https://reviewable.io/reviews/tailscale/tailscale/551)

### CODE REVIEW

File: tstest/natlab/nat.go [hunk 1]
```diff
@@ -0,0 +1,235 @@
...
+	return time.Now()
+}
+
+func (n *NAT44) initLocked() {
+	if n.byLAN == nil {
+		n.byLAN = map[natKey]*mapping{}
+		n.byWAN = map[netaddr.IPPort]*mapping{}
+	}
+	if n.MappingTimeout == 0 {
+		n.MappingTimeout = DefaultMappingTimeout
```
Comment (bradfitz) [inline tstest/natlab/nat.go:120]: I'd rather you didn't modify the public field and instead made an unexported accessor to get this value.

File: tstest/natlab/nat.go [hunk 2]
```diff
@@ -0,0 +1,235 @@
...
+	// direction.
+	Firewall PacketHandler
+	// TimeNow is a function that returns the current time. If
+	// nil, time.Now is used.
+	TimeNow func() time.Time
+
+	// inject, if not nil, will be invoked instead of Machine.Inject
+	// to inject NATed packets into the network. It is used for tests
+	// only.
+	inject func(p *Packet) error
```
Comment (bradfitz) [inline tstest/natlab/nat.go:100]: can drop the name `p` here. It's not used in docs.

File: tstest/natlab/nat.go [hunk 3]
```diff
@@ -0,0 +1,235 @@
...
+		panic(fmt.Sprintf("unknown NAT type %v", t))
+	}
+	return k
+}
+
+// DefaultMappingTimeout is the default timeout for a NAT mapping.
+const DefaultMappingTimeout = 30 * time.Second
+
+// NAT44 implements an IPv4-to-IPv4 source NAT (SNAT) translator, with
+// optional builtin firewall.
```
Comment (bradfitz) [inline tstest/natlab/nat.go:73]: It's not obvious to me how the NAT and firewall are wired together such that outbound NAT entries mappings are allowed back in.

File: tstest/natlab/nat.go [hunk 4]
```diff
@@ -0,0 +1,235 @@
...
+	"net"
+	"sync"
+	"time"
+
+	"inet.af/netaddr"
+)
+
+// mapping is the state of an allocated NAT session.
+type mapping struct {
+	// both tuples are from the POV of outbound packets, i.e. src is
```
[the remark this answers is not in the document]
Comment (danderson) [reply tstest/natlab/nat.go:19]: Comment was just outdated. Removed.

File: tstest/natlab/nat.go (hunk 3 above)
Comment (danderson) [reply tstest/natlab/nat.go:78]: The NAT logic invokes the firewall hook before NAT translation in the outbound direction, and after NAT translation in the inbound direction. So, the firewall handler always sees src=<lan addr>,dst=<remote addr>, as if the NAT didn't exist. That lets the NAT and firewall exist mostly orthogonally from each other.

Where should I put this in the docs?

**Pull request merged**